src/vivarium/framework/resource/resource.py

```diff
@@ -3,22 +3,45 @@
 from dataclasses import dataclass
 from typing import TYPE_CHECKING
 
+from vivarium.framework.lifecycle import LifeCycleError
+
 if TYPE_CHECKING:
     from vivarium import Component
     from vivarium.manager import Manager
 
 
-@dataclass
 class Resource:
     """A generic resource representing a node in the dependency graph."""
 
-    resource_type: str
-    """The type of the resource."""
-    name: str
-    """The name of the resource."""
-    # TODO [MIC-6433]: all resources should have a component
-    component: Component | Manager | None
-    """The component that creates the resource."""
+    def __init__(
+        self, resource_type: str, name: str, component: Component | Manager | None
+    ) -> None:
+        """Create a new resource.
+
+        Parameters
+        ----------
+        resource_type
+            The type of the resource.
+        name
+            The name of the resource.
+        component
+            The component or manager that produces this resource.
+        """
+        self.resource_type = resource_type
+        """The type of the resource."""
+        self.name = name
+        """The name of the resource."""
+        self._component = component
+        """The component that creates the resource. Can be None if not yet set."""
+
+    @property
+    def component(self) -> Component | Manager:
+        """The component that creates the resource."""
+        if self._component is None:
```

Comment:
I do not understand the use case where we hit this condition. The question I have though is why do we need to raise the LifeCycleError? Is it so we have an informative error? If component is now required when creating resources it seems like it should error if it is None then?